File: TEO/Mav.UI/Mav.Ui/src/app/services/i-user-assessment-section.service.ts
import { Observable } from 'rxjs/Observable';
import { GenericResult } from '../models/generic.result.model';
import { Paginate } from '../models/paginated.items';
import { UserAssessmentSectionSearchModel } from '../models/search/user.assessment.section.search.model';
import { UserAssessmentSectionListingModel } from '../models/search/user.assessment.section.listing.model';
import { SubmitUserAssessmentSectionViewModel } from '../models/submit.user.assessment.section.view.model';
import { UpdateUserAssessmentSectionViewModel } from '../models/update.user.assessment.section.view.model';
import { AddUserAssessmentSectionViewModel } from '../models/add.user.assessment.section.view.model';
import { AssessmentSectionTotals } from '../models/assessment.section.totals';
import { AssignQuestionItem } from '../models/assign.question.item';
import { ExamSectionStepViewModel } from '../models/exam-section-step.view.model';
import { UpdateUserSectionStatsViewModel } from '../models/update-user-section-stats.view.model';

export interface IUserAssessmentSectionService {

    getUserAssessmentSectionById<T>(id: number): Observable<T>;

    getUserAssessmentSectionQuestions(id: number): Observable<AssignQuestionItem[]>;

    getUserAssessmentSectionListByUserId(filter: UserAssessmentSectionSearchModel): Observable<Paginate<UserAssessmentSectionListingModel>>;

    getUserAssessmentSectionTotals(id: number): Observable<AssessmentSectionTotals>;

    lookupNextSectionStep(id: number): Observable<ExamSectionStepViewModel>;

    createUserAssessmentSection(newItem: AddUserAssessmentSectionViewModel): Observable<GenericResult<number>>;

    updateUserAssessmentSection(item: UpdateUserAssessmentSectionViewModel): Observable<GenericResult<number>>;

    submitUserAssessmentSection(item: SubmitUserAssessmentSectionViewModel): Observable<GenericResult<number>>;

    restartUserAssessmentSection(id: number): Observable<GenericResult<number>>;

    updateUserSectionStats(userSection: UpdateUserSectionStatsViewModel): Observable<GenericResult<number>>;

}

File: TEO/Mav.Services/Abstract/IUserAssessmentSectionService.cs
using Mav.Common.Models;
using Mav.Data.Repositories;
using Mav.Models.SearchModels;
using Mav.Models.ViewModels;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace Mav.Services.Abstract
{
    public interface IUserAssessmentSectionService
    {
        Task<T> GetById<T>(int id);

        Task<List<AssignQuestionItem>> GetUserAssessmentSectionQuestions(int id);

        Task<IPaginate<UserAssessmentSectionListingModel>> GetListByUserId(UserAssessmentSectionSearchModel vm);

        Task<AssessmentSectionTotals> GetUserAssessmentSectionTotals(int id);

        Task<List<AssessmentSectionSummary>> GetSectionSummaryDetails(int id);

        Task<int> GetRecentScore(int id);

        Task<ExamSectionStepViewModel> LookupNextSectionStep(int id);

        Task<GenericResult<T>> Add<T>(AddUserAssessmentSectionViewModel vm);

        Task<GenericResult<T>> Restart<T>(int id);

        Task<GenericResult<T>> Update<T>(UpdateUserAssessmentSectionViewModel vm);

        Task<GenericResult<T>> Submit<T>(SubmitUserAssessmentSectionViewModel vm);

        Task<AssignSectionQuestionsViewModel> AssignSectionQuestions(int sectionID);

        Task<GenericResult<T>> UpdateUserSectionStats<T>(UpdateUserSectionStatsViewModel vm);
    }
}
